File: Scripts/03-Figures/Run_All_Main_Text_Figures.py
#!/usr/bin/env python3
"""Run the R and Python scripts for CARB report Figures 2–17 in article order.

The runner can be launched with any Python 3 installation. Before plotting, it
finds a Python interpreter that has the required plotting packages. Set the
FIGURE_PYTHON environment variable to override automatic discovery.
"""

from pathlib import Path
import argparse
import os
import shutil
import subprocess
import sys

HERE = Path(__file__).resolve().parent
PROJECT_ROOT = HERE.parents[1]
FIGURE_SCRIPTS = Path("Scripts") / "03-Figures"
RSCRIPT = os.environ.get("RSCRIPT", "Rscript")
PYTHON_PACKAGES = ("matplotlib", "numpy", "pandas")
R_PACKAGES = (
    "dplyr", "geofacet", "ggplot2", "patchwork", "readr", "readxl",
    "scales", "stringr", "tidyr",
)


def python_has_packages(executable):
    """Return True when an interpreter can locate every plotting dependency."""
    package_list = repr(PYTHON_PACKAGES)
    check_code = (
        "import importlib.util, sys; "
        f"packages={package_list}; "
        "missing=[p for p in packages if importlib.util.find_spec(p) is None]; "
        "sys.exit(1 if missing else 0)"
    )
    try:
        result = subprocess.run(
            [str(executable), "-c", check_code],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            check=False,
        )
    except OSError:
        return False
    return result.returncode == 0


def python_candidates():
    """Yield unique Python executables, preferring an explicit override."""
    override = os.environ.get("FIGURE_PYTHON")
    if override:
        yield Path(override).expanduser()
        return

    seen = set()
    candidates = [Path(sys.executable)]
    executable_names = ("python3", "python", "python.exe")
    for directory in os.environ.get("PATH", "").split(os.pathsep):
        if not directory:
            continue
        for name in executable_names:
            candidates.append(Path(directory) / name)

    conda_prefix = os.environ.get("CONDA_PREFIX")
    if conda_prefix:
        candidates.extend(
            [Path(conda_prefix) / "bin" / "python3", Path(conda_prefix) / "python.exe"]
        )

    for candidate in candidates:
        if not candidate.is_file() or not os.access(candidate, os.X_OK):
            continue
        resolved = str(candidate.resolve())
        if resolved in seen:
            continue
        seen.add(resolved)
        yield Path(resolved)


def select_figure_python():
    override = os.environ.get("FIGURE_PYTHON")
    for candidate in python_candidates():
        if python_has_packages(candidate):
            return candidate

    package_text = " ".join(PYTHON_PACKAGES)
    if override:
        raise SystemExit(
            f"FIGURE_PYTHON={override!r} does not provide all required packages: "
            f"{package_text}.\nInstall them with:\n  {override} -m pip install {package_text}"
        )
    raise SystemExit(
        "No Python interpreter with all required plotting packages was found.\n"
        f"Required packages: {package_text}\n"
        "Install them in a Python environment, then set FIGURE_PYTHON to that "
        "interpreter."
    )


def check_r_environment():
    rscript_path = shutil.which(RSCRIPT)
    if rscript_path is None:
        raise SystemExit(
            f"Rscript executable {RSCRIPT!r} was not found. Set RSCRIPT to its path."
        )

    package_vector = ",".join(f'"{package}"' for package in R_PACKAGES)
    check_code = (
        f"required <- c({package_vector}); "
        "missing <- required[!vapply(required, requireNamespace, logical(1), quietly=TRUE)]; "
        "if (length(missing)) {"
        "cat('Missing R packages:', paste(missing, collapse=', '), '\\n'); quit(status=2)}"
    )
    result = subprocess.run(
        [rscript_path, "-e", check_code],
        cwd=PROJECT_ROOT,
        check=False,
    )
    if result.returncode != 0:
        raise SystemExit(
            "Install the missing R packages before running the figure pipeline."
        )
    return rscript_path


parser = argparse.ArgumentParser(description=__doc__)
parser.add_argument(
    "--check",
    action="store_true",
    help="Check Python and R plotting dependencies without generating figures.",
)
args = parser.parse_args()

FIGURE_PYTHON = select_figure_python()
RSCRIPT = check_r_environment()
print(f"Figure Python: {FIGURE_PYTHON}", flush=True)
print(f"Rscript: {RSCRIPT}", flush=True)
if args.check:
    print("All figure dependencies are available.", flush=True)
    raise SystemExit(0)

commands = [
    [RSCRIPT, FIGURE_SCRIPTS / "Fig02_EV_Share_New_LDV_Sales.R"],
    [RSCRIPT, FIGURE_SCRIPTS / "Fig03_Vehicle_Survival_Curves.R"],
    [RSCRIPT, FIGURE_SCRIPTS / "Fig04_North_American_New_Vehicle_Sales.R"],
    [RSCRIPT, FIGURE_SCRIPTS / "Fig05_California_LDV_Sales_and_Retirements.R"],
    [RSCRIPT, FIGURE_SCRIPTS / "Fig06_North_American_Vehicle_Retirements.R"],
    [RSCRIPT, FIGURE_SCRIPTS / "Fig07_North_American_Cumulative_Battery_Retirements.R"],
    [FIGURE_PYTHON, FIGURE_SCRIPTS / "Fig08_Regional_Supply_Chain_2050.py"],
    [FIGURE_PYTHON, FIGURE_SCRIPTS / "Fig09_Country_Supply_Chain_2050.py"],
    [FIGURE_PYTHON, FIGURE_SCRIPTS / "Fig10_North_American_Flows_Over_Time.py"],
    [FIGURE_PYTHON, FIGURE_SCRIPTS / "Fig11_Annual_Recycling_Capacity_Deficit.py", "--annual"],
    [FIGURE_PYTHON, FIGURE_SCRIPTS / "Fig12_Recoverable_Minerals.py"],
    [FIGURE_PYTHON, FIGURE_SCRIPTS / "Fig13_Cumulative_MRR_Losses_Through_2035.py"],
    [FIGURE_PYTHON, FIGURE_SCRIPTS / "Fig14_Maximum_RCS_North_America.py"],
    [FIGURE_PYTHON, FIGURE_SCRIPTS / "Supporting" / "RCS_Geographic_Analysis.py"],
    [FIGURE_PYTHON, FIGURE_SCRIPTS / "Supporting" / "Prepare_Fig17_Additional_Recycling_Capacity_Data.py"],
    [FIGURE_PYTHON, FIGURE_SCRIPTS / "Fig17_Additional_Recycling_Capacity_Required.py"],
]

environment = os.environ.copy()
environment["RCS_MAIN_FIGURE"] = "both"
for command in commands:
    print(f"Running {Path(command[1]).name}", flush=True)
    subprocess.run(
        [str(item) for item in command],
        check=True,
        cwd=PROJECT_ROOT,
        env=environment,
    )

File: Scripts/03-Figures/Run_All_Main_Text_Figures.R
#!/usr/bin/env Rscript

## Run CARB report Figures 2–17 from a single R entry point.
##
## Figures 2–7 are drawn by R scripts. Figures 8–17 retain their established
## Python/matplotlib implementations, but Python discovery, dependency checks,
## execution order, and error handling are managed here in R.

required_r_packages <- c(
  "dplyr", "geofacet", "ggplot2", "patchwork", "readr", "readxl",
  "scales", "stringr", "tidyr"
)
required_python_packages <- c("matplotlib", "numpy", "pandas")

find_this_script <- function() {
  file_arg <- grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE)
  if (length(file_arg)) {
    return(normalizePath(sub("^--file=", "", file_arg[[1]]), mustWork = TRUE))
  }

  frame_files <- vapply(
    sys.frames(),
    function(frame) {
      value <- get0("ofile", envir = frame, inherits = FALSE, ifnotfound = "")
      if (is.null(value)) "" else as.character(value)
    },
    character(1)
  )
  frame_files <- frame_files[nzchar(frame_files)]
  if (length(frame_files)) {
    return(normalizePath(tail(frame_files, 1), mustWork = TRUE))
  }

  fallback <- file.path(
    getwd(), "Scripts", "03-Figures", "Run_All_Main_Text_Figures.R"
  )
  normalizePath(fallback, mustWork = TRUE)
}

python_has_packages <- function(python) {
  package_literal <- paste0(
    "[", paste(sprintf("'%s'", required_python_packages), collapse = ","), "]"
  )
  check_code <- paste0(
    "import importlib.util,sys; packages=", package_literal, "; ",
    "missing=[p for p in packages if importlib.util.find_spec(p) is None]; ",
    "sys.exit(1 if missing else 0)"
  )
  status <- suppressWarnings(
    system2(
      python,
      args = c("-c", shQuote(check_code)),
      stdout = FALSE,
      stderr = FALSE
    )
  )
  is.numeric(status) && length(status) == 1 && status == 0
}

python_candidates <- function() {
  override <- Sys.getenv("FIGURE_PYTHON", unset = "")
  if (nzchar(override)) return(path.expand(override))

  path_dirs <- strsplit(
    Sys.getenv("PATH"), .Platform$path.sep, fixed = TRUE
  )[[1]]
  executable_names <- if (.Platform$OS.type == "windows") {
    c("python.exe", "python3.exe")
  } else {
    c("python3", "python")
  }

  candidates <- as.vector(outer(path_dirs, executable_names, file.path))
  conda_prefix <- Sys.getenv("CONDA_PREFIX", unset = "")
  if (nzchar(conda_prefix)) {
    candidates <- c(
      candidates,
      file.path(conda_prefix, "bin", "python3"),
      file.path(conda_prefix, "bin", "python"),
      file.path(conda_prefix, "python.exe")
    )
  }

  candidates <- c(
    candidates,
    path.expand("~/anaconda3/bin/python3"),
    path.expand("~/miniconda3/bin/python3"),
    "/opt/anaconda3/bin/python3"
  )
  candidates <- unique(candidates[file.exists(candidates)])
  unique(normalizePath(candidates, mustWork = TRUE))
}

select_python <- function() {
  override <- Sys.getenv("FIGURE_PYTHON", unset = "")
  candidates <- python_candidates()
  for (candidate in candidates) {
    if (python_has_packages(candidate)) return(candidate)
  }

  packages <- paste(required_python_packages, collapse = " ")
  if (nzchar(override)) {
    stop(
      "FIGURE_PYTHON does not contain all required packages: ", packages,
      "\nInstall them with:\n  ", override, " -m pip install ", packages,
      call. = FALSE
    )
  }
  stop(
    "No Python installation with all required figure packages was found.\n",
    "Required packages: ", packages, "\n",
    "Install them and optionally set FIGURE_PYTHON to that Python executable.",
    call. = FALSE
  )
}

run_command <- function(executable, arguments, label, environment = character()) {
  message("Running ", label)
  status <- system2(
    executable,
    args = arguments,
    stdout = "",
    stderr = "",
    env = environment
  )
  if (!is.numeric(status) || length(status) != 1 || status != 0) {
    stop(label, " failed with exit status ", status, ".", call. = FALSE)
  }
}

main <- function() {
  script_path <- find_this_script()
  project_root <- normalizePath(
    file.path(dirname(script_path), "..", ".."), mustWork = TRUE
  )
  old_wd <- setwd(project_root)
  on.exit(setwd(old_wd), add = TRUE)

  missing_r <- required_r_packages[
    !vapply(required_r_packages, requireNamespace, logical(1), quietly = TRUE)
  ]
  if (length(missing_r)) {
    stop(
      "Missing R packages: ", paste(missing_r, collapse = ", "),
      "\nInstall them before running the figure pipeline.",
      call. = FALSE
    )
  }

  python <- select_python()
  rscript <- file.path(R.home("bin"), "Rscript")
  if (!file.exists(rscript)) stop("Rscript was not found.", call. = FALSE)

  message("Project root: ", project_root)
  message("Rscript: ", rscript)
  message("Figure Python: ", python)

  check_only <- "--check" %in% commandArgs(trailingOnly = TRUE)
  if (check_only) {
    message("All figure dependencies are available.")
    return(invisible(TRUE))
  }

  figure_dir <- file.path("Scripts", "03-Figures")
  r_scripts <- c(
    "Fig02_EV_Share_New_LDV_Sales.R",
    "Fig03_Vehicle_Survival_Curves.R",
    "Fig04_North_American_New_Vehicle_Sales.R",
    "Fig05_California_LDV_Sales_and_Retirements.R",
    "Fig06_North_American_Vehicle_Retirements.R",
    "Fig07_North_American_Cumulative_Battery_Retirements.R"
  )
  python_scripts <- list(
    c("Fig08_Regional_Supply_Chain_2050.py"),
    c("Fig09_Country_Supply_Chain_2050.py"),
    c("Fig10_North_American_Flows_Over_Time.py"),
    c("Fig11_Annual_Recycling_Capacity_Deficit.py", "--annual"),
    c("Fig12_Recoverable_Minerals.py"),
    c("Fig13_Cumulative_MRR_Losses_Through_2035.py"),
    c("Fig14_Maximum_RCS_North_America.py"),
    c("Supporting/RCS_Geographic_Analysis.py"),
    c("Supporting/Prepare_Fig17_Additional_Recycling_Capacity_Data.py"),
    c("Fig17_Additional_Recycling_Capacity_Required.py")
  )

  for (script in r_scripts) {
    run_command(
      rscript,
      file.path(figure_dir, script),
      script
    )
  }

  matplotlib_cache <- file.path(tempdir(), "matplotlib")
  dir.create(matplotlib_cache, recursive = TRUE, showWarnings = FALSE)
  python_environment <- c(
    "RCS_MAIN_FIGURE=both",
    paste0("MPLCONFIGDIR=", matplotlib_cache)
  )
  for (script_args in python_scripts) {
    script <- script_args[[1]]
    arguments <- c(file.path(figure_dir, script), script_args[-1])
    run_command(
      python,
      arguments,
      basename(script),
      environment = python_environment
    )
  }

  message("All main-text figures completed successfully.")
  message("Figure data folder: ", file.path(project_root, "Results", "Data"))
  message("Final PNG folder: ", file.path(project_root, "Results", "Figures"))
  invisible(TRUE)
}

main()
